src/main/scala/com/fulcrumgenomics/bam/UpdateReadGroups.scala

```diff
@@ -0,0 +1,152 @@
+      |`ReplaceSamHeader` instead as in this case only the header needs modification. If all read groups are to be
+      |assigned to one read group, it is recommended to use Picard's `AddOrReplaceReadGroups`.  Nonetheless, if the read
+      |group identifier also needs to be changed, use this tool.
+      |
+      |Each read group in the input file will be mapped to one and only one new read group identifier, unless
+      |`ignoreMissingReadGroups` is set.  A SAM header file should be given with the new read groups and the ID field
+      |foreach read group containing the new read group identifier.  An additional attribute ("FR") should be provided
+      |that gives the original read group identifier ("ID") to which this new read group corresponds.
+      |
+      |If `keepReadGroupAttributes` is true, then any read group attribute not replaced will be kept in the new read
+      |group.  Otherwise, only the attributes in the provided SAM header file will be used.
+    """,
+  group = ClpGroups.SamOrBam)
+class UpdateReadGroups
+(@arg(flag = "i", doc = "Input BAM file.") val input: PathToBam,
+ @arg(flag = "o", doc = "Output BAM file.") val output: PathToBam,
+ @arg(flag = "r", doc = "A SAM header file with the replacement read groups (see detailed usage).") val readGroupsFile: PathToBam,
+ @arg(flag = "k", doc = "Keep all read group attributes that are not replaced.") val keepReadGroupAttributes: Boolean = false,
+ @arg(flag = "g", doc = "If true ignore read groups not found in the replacement header, otherwise throw an error.") val ignoreMissingReadGroups: Boolean = false
+) extends FgBioTool with LazyLogging {
+
+  private val fromReadGroupKey = "FR"
+
+  Io.assertReadable(input)
+  Io.assertReadable(readGroupsFile)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    val progress: ProgressLogger = new ProgressLogger(logger)
+
+    val in: SamReader = SamReaderFactory.make.open(input.toFile)
+    val toReadGroupsMap = getReadGroupMap(readGroupsFile, in.getFileHeader)
+    val outHeader = in.getFileHeader.clone()
+    outHeader.setReadGroups(toReadGroupsMap.values.toList.sortBy(_.getId))
+    val out: SAMFileWriter = new SAMFileWriterFactory().makeSAMOrBAMWriter(outHeader, true, output.toFile)
+
+    // main loop
+    in.foreach { record =>
+      updateRecord(record, outHeader, toReadGroupsMap)
+      out.addAlignment(record)
+      progress.record(record)
+    }
+    logger.info(s"Processed ${progress.getCount} in ${progress.getElapsedSeconds} seconds")
+
+    CloserUtil.close(in)
+    out.close()
+  }
+
+  private def updateRecord(record: SAMRecord,
+                           header: SAMFileHeader,
+                           readGroupMap: Map[String, SAMReadGroupRecord]): SAMRecord = {
+    val fromReadGroup = record.getReadGroup match {
+      case null => throw new IllegalStateException(s"Record '${record.getReadName}' has no read group")
```

Comment:
Since this is in a FgBioTool you can use `fail()` here instead which will render the message without printing out the stack trace.